### PR TITLE
ENH: feather support in the pandas IO api

### LINKED ISSUE #13092: ENH: pd.read_feather/to_feather
conda packages are now available [here](https://github.com/wesm/feather/issues/134)

so this should be straightforward. a couple of things to note.
- should support the `LocalPath/PathLib` objects (easy as we do this in other `read_*`)
- non-supported on windows ATM. need to skip for testing. `feather-format` will just raise an `ImportError` for now.
- catch and re-raise for non-supported things, mainly object dtypes that are not all strings. Alternatively / maybe better to introspect these columns (via `lib.infer_dtype`) to prevent in the first place
- test unicode support / encoding (not sure supported as of yet)
- biggest issue is how / if to support `Series` and a `DataFrame` with a non-default index. As we need a facility to write/read meta-data within feather [see here](https://github.com/wesm/feather/issues/74). Both these can simply `.reset_index()`, but the problem is they can be stored but we don't know on read-back what to `.set_index()`. Can just raise for now.

### PR DESCRIPTION
**Pull request by jreback**
closes #13092

### CODE REVIEW

File: pandas/io/feather_format.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                         "for the index; you can .reset_index()"
+                         "to make the index into column(s)")
+
+    if isinstance(df.columns, MultiIndex):
+        raise ValueError("feather does not serializing a MultiIndex "
+                         "for the columns")
+
+    # validate that we do not have any non-string object dtypes
+    # as these 'work', but will not properly de-serialize
+    valid_types = set(['string', 'unicode'])
```
Comment (jreback) [inline pandas/io/feather_format.py]: this provides a check for this serialization issue: https://github.com/wesm/feather/issues/240

**Comment by jreback** [comment]
cc @wesm

File: pandas/io/feather_format.py (hunk 1 above)
Comment (MaximilianR) [inline pandas/io/feather_format.py]: `{'string', 'unicode'}` works too FYI

Comment (jreback) [inline pandas/io/feather_format.py]: fixed....old habits :>

File: pandas/io/feather_format.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    # validate columns
+    # ----------------
+
+    # must have unique column names
+    if not df.columns.is_unique:
+        raise ValueError("feather does not support duplicate columns")
+
+    # must be a Index
+    if isinstance(df.columns, MultiIndex):
+        raise ValueError("feather does not serializing a MultiIndex "
```
Comment (wesm) [inline pandas/io/feather_format.py]: support

File: pandas/io/feather_format.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    # must have value column names (strings only)
+    if df.columns.inferred_type not in valid_types:
+        raise ValueError("feather must have string column names")
+
+    # validate dtypes
+    # ---------------
+
+    # validate that we do not have any non-string object dtypes
+    # as these 'work', but will not properly de-serialize
+    objects = df.select_dtypes(include=['object'])
```
Comment (wesm) [inline pandas/io/feather_format.py]: does this copy?

File: pandas/io/feather_format.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    if len(set(dtypes) - valid_types):
+        invalid = DataFrame([[i, c, dtype] for i, (c, dtype) in
+                             enumerate(zip(objects.columns, dtypes))])
+        invalid.columns = ['ncolumn', 'column', 'inferred_dtype']
+        invalid = invalid[~invalid.inferred_dtype.isin(list(valid_types))]
+
+        msg = ("The following columns are not supported to serialize "
+               "to the feather-format:\n\n"
+               "{}".format(invalid.to_string()))
+        raise ValueError(msg)
```
Comment (wesm) [inline pandas/io/feather_format.py]: I think this whole function should be pushed down into Feather itself, and the dtype inference can be handled there too

Comment (wesm) [inline pandas/io/feather_format.py]: Same goes for the unit tests...

File: pandas/io/feather_format.py (hunk 2 above)
Comment (jreback) [inline pandas/io/feather_format.py]: fixed

File: pandas/io/feather_format.py (hunk 3 above)
Comment (jreback) [inline pandas/io/feather_format.py]: yes, fixed another way

File: pandas/io/feather_format.py (hunk 4 above)
Comment (jreback) [inline pandas/io/feather_format.py]: yep

**Comment by jorisvandenbossche** [comment]
Shall we target this for 0.20?

**Comment by jreback** [comment]
some of these checks can be obviated by: https://github.com/wesm/feather/pull/244 & https://github.com/wesm/feather/pull/243

**Comment by jreback** [comment]
this is ready to go.

@jorisvandenbossche 
@wesm

File: doc/source/io.rst [hunk 5]
```diff
@@ -4135,6 +4137,63 @@ object). This cannot be changed after table creation.
...
+.. ipython:: python
+
+   df.to_feather('example.fth)
+
+Read from a feather file.
+
+.. ipython:: python
+
+   pd.read_feather('example.fth')
+
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: Maybe also show the dtypes? (so you see it is preserverd automatically)

File: doc/source/api.rst [hunk 6]
```diff
@@ -83,6 +83,14 @@ HDFStore: PyTables (HDF5)
    HDFStore.get
    HDFStore.select
 
+Feather
+~~~~~~~
+
+.. autosummary::
+   :toctree: generated/
+
+   read_feather
```
Comment (jorisvandenbossche) [inline doc/source/api.rst]: also `to_feather` in the dataframe section?

File: pandas/io/feather_format.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    feather = _try_import()
+    valid_types = {'string', 'unicode'}
+
+    # validate index
+    # --------------
+
+    # validate that we have only a default index
+    # raise on anything else as we don't serialize the index
+
+    if not isinstance(df.index, (RangeIndex, Int64Index)):
```
Comment (TomAugspurger) [inline pandas/io/feather_format.py]: Can this be simplified to `isinstance(df.index, Int64Index)`?

```python
In [6]: idx
Out[6]: RangeIndex(start=0, stop=10, step=1)

In [7]: isinstance(idx, pd.Int64Index)
Out[7]: True
```

Or is that not by design?

Comment (TomAugspurger) [inline pandas/io/feather_format.py]: I wonder if you should also check whether `df.index.name` is not None, and raise if it's set (or maybe warn?). I'm thinking about this too a bit in the JSON Table Schema stuff.

File: doc/source/io.rst [hunk 8]
```diff
@@ -4135,6 +4137,63 @@ object). This cannot be changed after table creation.
...
+Feather is designed to faithfully serialize and de-serialize DataFrames, supporting all of the pandas
+dtypes, including extension dtypes such as categorical and datetime with tz.
+
+Several caveats.
+
+- This is a newer library, and the format, though stable, is not guaranteed to be backward compatible
+  to the earlier versions.
+- The format will NOT write an ``Index``, or ``MultiIndex`` for the ``DataFrame`` and will raise an
+  error if a non-default one is provided. You can simply ``.reset_index()`` in order to store the index.
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: Additional point: Non-string column names ?

File: pandas/io/feather_format.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
+""" feather-format compat """
```
Comment (jorisvandenbossche) [inline pandas/io/feather_format.py]: Very minor, but can we call this file just `feather.py` ?

File: pandas/io/feather_format.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    # we need to import on first use
+
+    try:
+        import feather
+    except ImportError:
+
+        # give a nice error message
+        raise ImportError("the feather-format library is not installed\n"
+                          "you can install via conda\n"
+                          "conda install feather-format -c conda-forge")
```
Comment (jorisvandenbossche) [inline pandas/io/feather_format.py]: or pip?

File: pandas/io/feather_format.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    valid_types = {'string', 'unicode'}
+
+    # validate index
+    # --------------
+
+    # validate that we have only a default index
+    # raise on anything else as we don't serialize the index
+
+    if not isinstance(df.index, (RangeIndex, Int64Index)):
+        raise ValueError("feather does not serializing {} "
```
Comment (jorisvandenbossche) [inline pandas/io/feather_format.py]: either "does not support serializing" or "does not serialize"

File: pandas/io/feather_format.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+                         "for the index; you can .reset_index()"
+                         "to make the index into column(s)".format(
+                             type(df.index)))
+
+    if not df.index.equals(RangeIndex.from_range(range(len(df)))):
+        raise ValueError("feather does not serializing a non-default index "
```
Comment (jorisvandenbossche) [inline pandas/io/feather_format.py]: the same here

File: doc/source/io.rst (hunk 8 above)
Comment (jorisvandenbossche) [inline doc/source/io.rst]: and duplicate column names

File: pandas/io/feather_format.py (hunk 9 above)
Comment (jreback) [inline pandas/io/feather_format.py]: yeah that is an annoying 'feature' in python!

File: doc/source/io.rst (hunk 8 above)
Comment (jreback) [inline doc/source/io.rst]: yep, these are raised automaticaly by feather now (as of 3.1)

File: pandas/io/feather_format.py (hunk 10 above)
Comment (jreback) [inline pandas/io/feather_format.py]: hmm, on linux I suppose this is ok (maybe mac), but certainly not on windows. 

@wesm ?

**Comment by jreback** [comment]
all comments addressed

**Comment by jreback** [comment]
any final comments. @wesm ?

going to merge in the next day or 2.

File: pandas/io/feather_format.py (hunk 10 above)
Comment (wesm) [inline pandas/io/feather_format.py]: It'll work on Windows if you have the right compiler setup. Can recommend both here I suppose

**Review by wesm — APPROVED** [review]
LGTM

**Comment by jreback** [comment]
ok added pip install instructions as well.

**Comment by jorisvandenbossche** [comment]
@jreback we should also add feather-format to the list of the doc building requirements (ci/requirements-2.7_DOC_BUILD.run)

**Comment by jreback** [comment]
yep will do

**Comment by jreback** [comment]
@jorisvandenbossche https://github.com/pandas-dev/pandas/commit/3e3434b21e3c135c3bd7649689d488380ea1ebc1

**Comment by jorisvandenbossche** [comment]
@jreback Seems to be working now! There was only still a small typo: https://github.com/pandas-dev/pandas/commit/975a440b08cce771e286de86706fa213d6479d45

**Comment by jorisvandenbossche** [comment]
@jreback @wesm A more stylistic question: Jeff now used `.fth` as the file extension, but I think I have mainly seen `.feather` (in blogposts about it). Best to be consistent here. @wesm do you prefer one of both?

**Comment by wesm** [comment]
We've been using `.feather`, maybe stick with that.

**Comment by jorisvandenbossche** [comment]
OK, done: https://github.com/pandas-dev/pandas/commit/e27b29697f0dcf9359f01a19edb2f20c6d728b6c

**Pull request closed without merging**